### PR TITLE
Expose an eeIsJitIntrinsic method for quickly determining if a method handle represents an intrinsic method

### LINKED ISSUE #51006: Regression in Crossgen2 Throughput
### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single - Microsoft.CodeAnalysis

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[CodeAnalysis](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20-%20Microsoft.CodeAnalysis.html>) | 1854.95 ms | 1921.10 ms | 1.04 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20-%20Microsoft.CodeAnalysis.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single - Microsoft.CodeAnalysis*'
```
<details>

#### Crossgen2 Throughput - Single - Microsoft.CodeAnalysis

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>
### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single - System.Net.Sockets

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[Sockets](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20-%20System.Net.Sockets.html>) | 775.46 ms | 786.31 ms | 1.01 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20-%20System.Net.Sockets.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single - System.Net.Sockets*'
```
<details>

#### Crossgen2 Throughput - Single - System.Net.Sockets

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single - System.Private.CoreLib

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[CoreLib](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20-%20System.Private.CoreLib.html>) | 2741.40 ms | 3018.09 ms | 1.10 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20-%20System.Private.CoreLib.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single - System.Private.CoreLib*'
```
<details>

#### Crossgen2 Throughput - Single - System.Private.CoreLib

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single Threaded - Single - System.Data.Common

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[Common](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20Threaded%20-%20Single%20-%20System.Data.Common.html>) | 2563.62 ms | 2740.17 ms | 1.07 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20Threaded%20-%20Single%20-%20System.Data.Common.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single Threaded - Single - System.Data.Common*'
```
<details>

#### Crossgen2 Throughput - Single Threaded - Single - System.Data.Common

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single - Microsoft.CodeAnalysis.VisualBasic

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[VisualBasic](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20-%20Microsoft.CodeAnalysis.VisualBasic.html>) | 3807.93 ms | 3969.92 ms | 1.04 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20-%20Microsoft.CodeAnalysis.VisualBasic.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single - Microsoft.CodeAnalysis.VisualBasic*'
```
<details>

#### Crossgen2 Throughput - Single - Microsoft.CodeAnalysis.VisualBasic

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single Threaded - Single - Microsoft.CodeAnalysis.CSharp

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[CSharp](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20Threaded%20-%20Single%20-%20Microsoft.CodeAnalysis.CSharp.html>) | 7879.24 ms | 8506.36 ms | 1.08 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20Threaded%20-%20Single%20-%20Microsoft.CodeAnalysis.CSharp.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single Threaded - Single - Microsoft.CodeAnalysis.CSharp*'
```
<details>

#### Crossgen2 Throughput - Single Threaded - Single - Microsoft.CodeAnalysis.CSharp

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single Threaded - Single - Microsoft.CodeAnalysis

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[CodeAnalysis](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20Threaded%20-%20Single%20-%20Microsoft.CodeAnalysis.html>) | 3577.61 ms | 3794.41 ms | 1.06 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20Threaded%20-%20Single%20-%20Microsoft.CodeAnalysis.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single Threaded - Single - Microsoft.CodeAnalysis*'
```
<details>

#### Crossgen2 Throughput - Single Threaded - Single - Microsoft.CodeAnalysis

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Composite - framework-r2r

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[Crossgen2 Throughput - Composite - framework-r2r](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Composite%20-%20framework-r2r.html>) | 32497.26 ms | 35029.59 ms | 1.08 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Composite%20-%20framework-r2r.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Composite - framework-r2r*'
```
<details>

#### Crossgen2 Throughput - Composite - framework-r2r

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single Threaded - Single - System.Net.WebProxy

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[WebProxy](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20Threaded%20-%20Single%20-%20System.Net.WebProxy.html>) | 661.70 ms | 667.57 ms | 1.01 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20Threaded%20-%20Single%20-%20System.Net.WebProxy.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single Threaded - Single - System.Net.WebProxy*'
```
<details>

#### Crossgen2 Throughput - Single Threaded - Single - System.Net.WebProxy

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single - System.Data.Common

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[Common](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20-%20System.Data.Common.html>) | 1268.30 ms | 1327.32 ms | 1.05 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20-%20System.Data.Common.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single - System.Data.Common*'
```
<details>

#### Crossgen2 Throughput - Single - System.Data.Common

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single - System.ServiceModel.Syndication

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[Syndication](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20-%20System.ServiceModel.Syndication.html>) | 726.85 ms | 732.04 ms | 1.01 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20-%20System.ServiceModel.Syndication.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single - System.ServiceModel.Syndication*'
```
<details>

#### Crossgen2 Throughput - Single - System.ServiceModel.Syndication

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Diff | [Diff](https://github.com/dotnet/runtime/compare/a3e0894f38b1ab558653e7c1b7322a8c4db92af1...bdd7ca0d78ce86c2dc0ab24df42a09ef514a771c)

### Regressions in Crossgen2 Throughput - Single Threaded - Single - System.Net.Sockets

Benchmark | Baseline | Test | Test/Base | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | --
[Sockets](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_RunKind=crossgen_scenarios/Crossgen2%20Throughput%20-%20Single%20Threaded%20-%20Single%20-%20System.Net.Sockets.html>) | 1106.07 ms | 1149.64 ms | 1.04 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/04_08_2021/refs/heads/main_x64_ubuntu%2018.04_RunKind=crossgen_scenarios_Regression/Crossgen2%20Throughput%20-%20Single%20Threaded%20-%20Single%20-%20System.Net.Sockets.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Crossgen2 Throughput - Single Threaded - Single - System.Net.Sockets*'
```
<details>

#### Crossgen2 Throughput - Single Threaded - Single - System.Net.Sockets

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### PR DESCRIPTION
**Pull request by tannergooding**
This should resolve https://github.com/dotnet/runtime/issues/51006

### CODE REVIEW

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 1]
```diff
@@ -858,6 +858,12 @@ private TypeSystemEntity entityFromContext(CORINFO_CONTEXT_STRUCT* contextStruct
             return (TypeSystemEntity)HandleToObject((IntPtr)((ulong)contextStruct & ~(ulong)CorInfoContextFlags.CORINFO_CONTEXTFLAGS_MASK));
         }
 
+        private bool isJitIntrinsic(CORINFO_METHOD_STRUCT_* ftn)
+        {
+            MethodDesc method = HandleToObject(ftn);
+            return method.IsIntrinsic;
```
Comment (tannergooding) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:864]: I think the only particular thing of note is this (as per https://github.com/dotnet/runtime/compare/main...tannergooding:isJitIntrinsic?expand=1#diff-132a77bcd3f74cf0e0b04fbccda246c97c91e40562d78cb01fff61cf69403573L872-L873) isn't fine grained enough to determine just `CORINFO_FLG_INTRINSIC` or `CORINFO_FLG_JIT_INTRINSIC`

**Comment by AndyAyersMS** [comment]
Did you measure crossgen2 throughput with this? You could probably also use crossgen1, it might be simpler.

**Comment by tannergooding** [comment]
> Did you measure crossgen2 throughput with this? You could probably also use crossgen1, it might be simpler.

Working on doing that now, I was just planning on running the same benchmark from https://github.com/dotnet/runtime/issues/51006

**Comment by tannergooding** [comment]
@AndyAyersMS, @jkotas, on my own box, I see the below.

### Crossgen2

Base:
```
    Crossgen2 Throughput - Single - System.Private.Corelib
    Metric               |Average        |Min            |Max
    ---------------------|---------------|---------------|---------------
    Process Time         |2195.994 ms    |2152.472 ms    |2241.161 ms
    Loading Interval     |478.729 ms     |473.132 ms     |488.765 ms
    Emitting Interval    |585.857 ms     |568.642 ms     |601.387 ms
    Jit Interval         |831.327 ms     |810.561 ms     |864.635 ms
    Compilation Interval |2143.333 ms    |2097.013 ms    |2188.326 ms
    Time on Thread       |11848.433 ms   |11634.819 ms   |12010.432 ms
```

Resolve tokens (+6.03% from Base):
```
    Crossgen2 Throughput - Single - System.Private.Corelib
    Metric               |Average        |Min            |Max
    ---------------------|---------------|---------------|---------------
    Process Time         |2328.495 ms    |2242.713 ms    |2375.665 ms
    Loading Interval     |494.870 ms     |474.765 ms     |510.811 ms
    Emitting Interval    |582.028 ms     |570.934 ms     |595.669 ms
    Jit Interval         |937.977 ms     |903.238 ms     |985.727 ms
    Compilation Interval |2274.020 ms    |2188.263 ms    |2322.155 ms
    Time on Thread       |13275.996 ms   |12798.085 ms   |13556.782 m
```

Jan's change (+2.24% from Base, -3.58% from Resolve tokens):
```
    Crossgen2 Throughput - Single - System.Private.Corelib
    Metric               |Average        |Min            |Max
    ---------------------|---------------|---------------|---------------
    Process Time         |2245.198 ms    |2207.725 ms    |2283.867 ms
    Loading Interval     |484.585 ms     |477.682 ms     |502.124 ms
    Emitting Interval    |569.878 ms     |557.664 ms     |585.911 ms
    Jit Interval         |905.695 ms     |872.238 ms     |917.978 ms
    Compilation Interval |2192.197 ms    |2155.402 ms    |2230.611 ms
    Time on Thread       |13092.984 ms   |12916.836 ms   |13276.031 ms
```

This change (+1.94% from Base, -3.86% from Resolve tokens, -0.29% from Jan's change):
```
    Crossgen2 Throughput - Single - System.Private.Corelib
    Metric               |Average        |Min            |Max
    ---------------------|---------------|---------------|---------------
    Process Time         |2238.630 ms    |2178.115 ms    |2301.655 ms
    Loading Interval     |479.965 ms     |473.764 ms     |485.390 ms
    Emitting Interval    |568.050 ms     |564.785 ms     |574.498 ms
    Jit Interval         |894.025 ms     |854.444 ms     |941.489 ms
    Compilation Interval |2186.992 ms    |2126.187 ms    |2249.631 ms
    Time on Thread       |11828.096 ms   |11491.262 ms   |12034.082 ms
```

Notably, its not clear why this change, which is doing a lot less work than `eeGetCallInfo` is such a small difference (nor why "maxTime" is worse).

**Comment by jkotas** [comment]
The "Time on Thread" has improved significantly with this change. Is it a sign that this change works well?

**Comment by AndyAyersMS** [comment]
Error bars in lab measurements suggest there is a few % variation in run times for crossgen2.

image

How many runs are you averaging?

**Comment by tannergooding** [comment]
> Is it a sign that this change works well?

@DrewScoggins might be able to give more insight into what each value is for.
My hypothesis is `Jit Interval` is the main one we want to be looking at here, given that these changes are in the JIT and should likely impact that.

There is a good bit of noise here as well, and multiple independent runs can result in some pretty variable results.

For example, in the change I just pushed which drops the `impResolveToken` call for `CEE_CONSTRAINED` (since we don't need it given we aren't calling `eeGetCallInfo` anymore), has higher `process time` and higher `time on thread`, but less `Jit interval`:

```
    Crossgen2 Throughput - Single - System.Private.Corelib
    Metric               |Average        |Min            |Max
    ---------------------|---------------|---------------|---------------
    Process Time         |2247.620 ms    |2192.448 ms    |2386.135 ms
    Loading Interval     |480.200 ms     |475.420 ms     |485.793 ms
    Emitting Interval    |587.992 ms     |562.316 ms     |660.547 ms
    Jit Interval         |870.072 ms     |846.114 ms     |920.709 ms
    Compilation Interval |2195.054 ms    |2140.109 ms    |2334.002 ms
    Time on Thread       |12126.867 ms   |11708.378 ms   |12998.084 ms
```

**Comment by tannergooding** [comment]
> How many runs are you averaging?

Console output indicates 1 warm up and 5 iterations, for all 4 (now 5) shared results.
The results for the crossgen tests are a bit different from the regular benchmarks.

**Comment by AndyAyersMS** [comment]
Given that we're making measurements where we don't have a good model for the underlying distribution, median may be a more useful thing to report than mean.

Historically for TP measurements the jit team would use crossgen1 running under PIN (or some other instruction counting technique), and there the run to run repeatability was quite good. So you might try doing that instead.

**Comment by tannergooding** [comment]
I can run some more benchmarks later this afternoon. I can collect some instruction counts using AMD uProf (`pin` largely only works on Intel).

I'll need to figure the command line the benchmark is passing to crossgen as part of that.

**Comment by DrewScoggins** [comment]
```
if compiletype == const.CROSSGEN2_SINGLEFILE:
            referencefilenames = ['System.*.dll', 'Microsoft.*.dll', 'netstandard.dll', 'mscorlib.dll']
            # single assembly filename: example.dll
            filename, ext = os.path.splitext(self.singlefile)
            outputdir = os.path.join(os.getcwd(), const.CROSSGENDIR)
            if not os.path.exists(outputdir):
                os.mkdir(outputdir)
            outputfile = os.path.join(outputdir, filename+'.ni'+ext )
            
            crossgen2args = [
                os.path.join(self.coreroot, self.singlefile),
                '-o', outputfile,
                '-O'
            ]

            for reffile in referencefilenames:
                crossgen2args.extend(['-r', os.path.join(self.coreroot, reffile)])
```

Here is the code that we use to generate the command line. Looks like we call crossgen2 with the path to the file we are crossgening, give an output file name and use the -O flag as well as -r for the different ref files.

**Comment by MichalStrehovsky** [comment]
> Error bars in lab measurements suggest there is a few % variation in run times for crossgen2.

For the Native AOT compiler (which is pretty much crossgen2), turning off Tiered JIT stabilized throughput by *a lot*. We actually also had a compilation throughput improvement with that so it's [just off by default there](https://github.com/dotnet/runtimelab/pull/750).

**Comment by tannergooding** [comment]
@AndyAyersMS, @jkotas

I'm seeing an average of:
* 22 retired instructions in "Base"
* 234 retired instructions in "Resolve tokens"
* 190 retired instructions in "Jan's change"
* 224 retired instructions in "This change"

It isn't clear to me why this is the case initially, since this should be doing less work now.
Are we aware of anything that might impact this, like changing the JITEEVersionIdentifier or modifying the JIT/EE interface invalidating something (PGO data, some cache, etc)

**Comment by tannergooding** [comment]
Rebased to resolve merge conflict

**Comment by jkotas** [comment]
What does the retired instructions number mean?

**Comment by tannergooding** [comment]
> What does the retired instructions number mean?

The number of instructions that were executed and correct (meaning this does not include speculatively executed instructions).
This is roughly equivalent to the instruction count output by the `pin` tool Andy mentioned.

In general, it isn't clear why calling `eeIsJitIntrinsic` is not cheaper than calling `eeGetCallInfo`, considering the latter does everything `eeIsJitIntrinsic` does, plus more.

**Comment by AndyAyersMS** [comment]
We usually just look at the overall numbers for the whole process, not try and attribute them to particular methods.

**Comment by tannergooding** [comment]
> We usually just look at the overall numbers for the whole process, not try and attribute them to particular methods.

Its lost in the noise of the overall process. The average for all 4 variations, measured across 5 runs each, is about 92000 retired instructions with not enough variation between them to see anything discernable in terms of hardware metrics that ETL or uProf collect.

**Comment by AndyAyersMS** [comment]
This small variation in instructions retired is seen across all 4 variants? (base, regression, fix1, fix2...)? If so then something is off in the measurement process.

I can try looking with PIN later and post what I see.

**Comment by AndyAyersMS** [comment]
Here's the full process pin icount data for crossgen SPC. Seems to track your crossgen2 data above fairly well.  

Looks like we're maybe 1.5% slower than we were before all this. I think we can live with that.
```
Baseline
Count 19254296972
Count 19256217769
Count 19253906826

Resolve Tokens
Count 20138271494
Count 20142288269
Count 20140098703

Jan's fix
Count 19988303504
Count 19988649133
Count 19990292972

This change
Count 19566027492
Count 19567077283
Count 19564959405
```

**Review by AndyAyersMS — APPROVED** [review]

**Review by jkotas — APPROVED** [review]

**Comment by tannergooding** [comment]
Thanks! I'll need to figure out why uProf wasn't reporting the right numbers on my end

**Pull request merged**